### ISSUE TITLE
Write tests for twisted.web.proxy

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11333]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#238|
|Type|enhancement|
|Created|2003-08-26 21:15:37Z|
|Branch|https://github.com/twisted/twisted/tree/proxy-tests-238|

<details><summary>Searchable metadata</summary>

```
trac-id__238 238
type__enhancement enhancement
reporter__itamarst itamarst
priority__highest highest
milestone__ 
branch__branches_proxy_tests_238 branches/proxy-tests-238
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1061932537000000 1061932537000000
changetime__1195597424000000 1195597424000000
version__ 
owner__ 
cc__itamarst cc__etrepum cc__jknight cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [21830]) Branching to 'proxy-tests-238'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [21831]) Look tests!

Refs [#11333](https://github.com/twisted/twisted/issues/11333)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml set owner to @therve|
|-|-|

Hooray tests! I love tests. Thank you so much for shining the light of unit tests onto this dark code.

`twisted/web/proxy.py`
 * Comment in `handleStatus` goes over 79 columns. Please wrap.
 * Docstring for `proxyClientFactoryClass` in `ReverseProxyRequest` goes over 79 columns. Please wrap.
 * Indentation in `ReverseProxyResource.render` is incorrect. Please change from:
 ```python
        clientFactory = self.proxyClientFactoryClass(request.method, rest,
                                     request.clientproto,
                                     request.getAllHeaders(),
                                     request.content.read(),
                                     request)
```

 to
 ```python
        clientFactory = self.proxyClientFactoryClass(
            request.method, rest, request.clientproto,
            request.getAllHeaders(), request.content.read(), request)
```


`twisted/web/test/test_proxy.py`
 * The strategy you use for stubbing the reactor in `TestableReverseProxyResource` isn't the best. Instead, you should pass in a fake reactor.
 * There's a lot of common code between `ReverseProxyResourceTestCase.test_render` and `ReverseProxyResourceTestCase.test_renderWithQuery`. It would be improved with a custom assert method that clearly communicates the intent of the various sub-assertions. Similarly, there's some set-up code that could be parametrized and extracted. I think that once you have some custom assertions, a lot of the other tests in this file could become much shorter.
 * `DummyFather` should be renamed to `DummyParent` and other references to 'father' changed to 'parent'.
 * `ProxyClientTestCase` also has two tests that look very similar and are kind of long. 
 * In `test_statusWithMessage`, the docstring says "correctly be forwarded", correct English is "be forwarded correctly", but...
 * Using the word 'correct' in a docstring for a test isn't very helpful. The reader already knows (or assumes) that you are testing for correct behaviour. It's generally better if you can describe what "correct" means.

Overall, the tests are each testing one thing, appear to be testing the right things and have clear docstrings -- congratulations.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [21839]) Process review comments.

Refs [#11333](https://github.com/twisted/twisted/issues/11333)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [21873]) Some comments.

Refs [#11333](https://github.com/twisted/twisted/issues/11333)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Thanks a lot for your review, that should be better.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

  * twisted/web/test/test_proxy.py
    * what's up with ```channel.callLater = clock.callLater```?  It seems extraneous.
    * ProxyClientTestCase._testDataForward should reverse the order of makeConnection and setting the transport - in case ProxyClient.makeConnection wants to do something that might require the transport to know its protocol.
  * twisted/web/proxy.py
    * in !ProxyRequest.process, ```if not 'host' in headers:``` is not as clear as ```if 'host' not in headers:```
    * !ProxyRequest could document its ```reactor``` attribute.
    * Ditto !ReverseProxyRequest and !ReverseProxyResource
    * It's not clear why the trailing / is stripped from the path in !ReverseProxyResource.!__init!__.  I guess it avoids duplicating the / when making a request... but why not just do what the user asked?  What happens if the path specified to !__init!__ ends with two /s?

Still no tests for ```POST```s, or more generally... requests with bytes in their request body, I think?  Not sure if you want to separate that from the rest of this or not.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [21882]) Process review comments.

Refs [#11333](https://github.com/twisted/twisted/issues/11333)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [21883]) Some test for requests with body.

Refs [#11333](https://github.com/twisted/twisted/issues/11333)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @therve|
|-|-|

The change to ```ReverseProxyResource.getChild``` isn't really tested, afaik - the isinstance assertion in test_getChild will pass without the change from using ```ReverseProxyResource``` to using ```self.__class__```.  Generally, I'm not sure this change really fits with the rest of the branch anyway.  Why is it a useful behavior?

The comment in ```ReverseProxyResource.__init__``` about path and the slash isn't quite clear to me.  Is it supposed to say _The slash will be added ..._?  I'm still not sure why removing the slash is necessary though.

Everything else looks good now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

(In [21904]) Process review comments.

Refs [#11333](https://github.com/twisted/twisted/issues/11333)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve commented|
|-|-|

(In [21905]) Add documentation.

Refs [#11333](https://github.com/twisted/twisted/issues/11333)
Refs [#2905](https://github.com/twisted/twisted/issues/2905)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve removed owner|
|-|-|

Just to be sure that the last docstring is OK.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @therve|
|-|-|

Looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set status to `closed`|
|-|-|

(In [21922]) Merge proxy-tests-238

Author: therve
Reviewers: exarkun, jml
Fixes [#11333](https://github.com/twisted/twisted/issues/11333)
Fixes [#2905](https://github.com/twisted/twisted/issues/2905)

Add tests to twisted.web.proxy.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
